### ISSUE TITLE
OpenSSL violates OCSP RFC by always assuming sha1/pre-existing EVP_MD

### ISSUE THREAD

**Issue by mlindner** [openssl/openssl#8561]
https://tools.ietf.org/html/rfc6960#section-4.3

```
4.3.  Mandatory and Optional Cryptographic Algorithms

   Clients that request OCSP services SHALL be capable of processing
   responses signed using RSA with SHA-256 (identified by the
   sha256WithRSAEncryption OID specified in [RFC4055]).  Clients SHOULD
   also be capable of processing responses signed using RSA with SHA-1
   (identified by the sha1WithRSAEncryption OID specified in [RFC3279])
   and the Digital Signature Algorithm (DSA) with SHA-1 (identified by
   the id-dsa-with-sha1 OID specified in [RFC3279]).  Clients MAY
   support other algorithms.
```

The current way OpenSSL works in 1.1.1 makes it impossible to follow the RFC while also using the API. `OCSP_resp_find`, `OCSP_resp_find_status` and family assume that you already have a `OCSP_CERTID` structure, however that is impossible because the digest hash algorithm is selected by the server. It used to be possible to extract the hash algorithm from the OCSP_SINGLERESP (iterating through them) via `OBJ_obj2nid(single->certId->hashAlgorithm->algorithm)` however this no longer appears to be possible.

**Comment by mlindner** [comment]
ping

**Comment by mlindner** [comment]
I can add a pull request that has rewritten/added OCSP functions that instead take in a X509* cert and X509* issuer and generates the OCSP_CERTID objects on the fly according to the algorithm located in the OCSP_SINGLERESP. Would this be acceptable?

**Comment by mlindner** [comment]
This is needed for 1.1.1 for correct functioning of OCSP.

**Comment by mattcaswell** [comment]
> I can add a pull request that has rewritten/added OCSP functions that instead take in a X509* cert and X509* issuer and generates the OCSP_CERTID objects on the fly according to the algorithm located in the OCSP_SINGLERESP. Would this be acceptable?

Can you suggest what the signatures for such functions might look like and how they would be used?

> It used to be possible to extract the hash algorithm from the OCSP_SINGLERESP (iterating through them) via OBJ_obj2nid(single->certId->hashAlgorithm->algorithm) however this no longer appears to be possible.

It is possible to get the certid from the OCSP_SINGLERESP using OCSP_SINGLERESP_get0_id, although there does not seem a way to get the hash algorithm from it (perhaps an accessor is missing here?)

**Comment by mlindner** [comment]
> It is possible to get the certid from the OCSP_SINGLERESP using OCSP_SINGLERESP_get0_id, although there does not seem a way to get the hash algorithm from it (perhaps an accessor is missing here?)

Yes that is what I would add and what I've modified locally (unfortunately do not yet have permission to submit code to OpenSSL project). This is needed in 1.1.1.

> Can you suggest what the signatures for such functions might look like and how they would be used?

They would basically be a replacement for `OCSP_resp_find_status` that have identical function signatures but instead take cert/cert_issuer pairs (of X509* type) instead of `OCSP_CERTID` and then uses the hash algorithm from every `OCSP_SINGLERESP` in the `OCSP_BASICRESP` to hash the cert/issuers to generate `OCSP_CERTID` and continues to attempt to do so until a matching generated OCSP_CERTID` entry is found or the `OCSP_BASICRESP` is exhausted.

**Comment by mlindner** [comment]
You can see this illustrated as a bug in several open source projects. For example in NGINX:

https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/event/ngx_event_openssl_stapling.c#L669-L683

OCSP_cert_to_id passed in with NULL for the EVP_MD (because it's unknown at this point) causes EVP_sha1() to be selected.

https://github.com/openssl/openssl/blob/cc6d92619fc3678817b2e09894683b40860563a7/crypto/ocsp/ocsp_lib.c#L29

If the response did not use sha1 as the hash algorithms (for example some codebases use sha256 by default for OCSP responses) then this NGINX OCSP code will fail.

**Comment by mattcaswell** [comment]
>> It is possible to get the certid from the OCSP_SINGLERESP using OCSP_SINGLERESP_get0_id, although there does not seem a way to get the hash algorithm from it (perhaps an accessor is missing here?)
>
> Yes that is what I would add and what I've modified locally (unfortunately do not yet have permission to submit code to OpenSSL project). This is needed in 1.1.1.
For example: const X509_ALGOR *OCSP_CERTID_get0_sigalg(const OCSP_CERTID *id);

Ah! It turns out there *is* an accessor for this:

````
 int OCSP_id_get0_info(ASN1_OCTET_STRING **piNameHash, ASN1_OBJECT **pmd,
                       ASN1_OCTET_STRING **pikeyHash,
                       ASN1_INTEGER **pserial, OCSP_CERTID *cid);
````
Documented here:
https://www.openssl.org/docs/man1.1.1/man3/OCSP_id_get0_info.html

Using that call you can get hold of the oid for the hash algorithm (in the `pmd` parameter) and then use `EVP_get_digestbyobj` to get hold of the EVP_MD object:

https://www.openssl.org/docs/man1.1.1/man3/EVP_get_digestbyobj.html

**Comment by mattcaswell** [comment]
> I can add a pull request that has rewritten/added OCSP functions that instead take in a X509* cert and X509* issuer and generates the OCSP_CERTID objects on the fly according to the algorithm located in the OCSP_SINGLERESP. Would this be acceptable?

I would welcome a PR for this. For backwards compat we must keep the old functions - so this would be "added" functionality rather than a replacement.

**Comment by mlindner** [comment]
> > > It is possible to get the certid from the OCSP_SINGLERESP using OCSP_SINGLERESP_get0_id, although there does not seem a way to get the hash algorithm from it (perhaps an accessor is missing here?)
> > 
> > 
> > Yes that is what I would add and what I've modified locally (unfortunately do not yet have permission to submit code to OpenSSL project). This is needed in 1.1.1.
> > For example: const X509_ALGOR *OCSP_CERTID_get0_sigalg(const OCSP_CERTID *id);
> 
> Ah! It turns out there _is_ an accessor for this:
> 
> ```
>  int OCSP_id_get0_info(ASN1_OCTET_STRING **piNameHash, ASN1_OBJECT **pmd,
>                        ASN1_OCTET_STRING **pikeyHash,
>                        ASN1_INTEGER **pserial, OCSP_CERTID *cid);
> ```
> Documented here:
> https://www.openssl.org/docs/man1.1.1/man3/OCSP_id_get0_info.html
> 
> Using that call you can get hold of the oid for the hash algorithm (in the `pmd` parameter) and then use `EVP_get_digestbyobj` to get hold of the EVP_MD object:
> 
> https://www.openssl.org/docs/man1.1.1/man3/EVP_get_digestbyobj.html

Ah great! That is very much misnamed and the documentation needs to be better as "pmd" and "hash OID" do not connect in my head to hash algorithm. I didn't understand what it was doing until I read the source (also the manual file doesn't have the normal formatting where it puts each parameter in "bold" lettering explaining what each is). Possibly some changes are needed there.

> I would welcome a PR for this. For backwards compat we must keep the old functions - so this would be "added" functionality rather than a replacement.

When I get the ability to I will submit these, can these be backported as well?

**Comment by mattcaswell** [comment]
> When I get the ability to I will submit these, can these be backported as well?

Unfortunately we don't allow new features in stable releases so these new functions wouldn't be eligible for backport.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
